### PR TITLE
raftstore: enhance raft snapshot metrics

### PR DESCRIPTION
**Pull request by hhwyt**
This pull request enhances metrics for the raft snapshot workflow, includes:
1.Use generate-kv, generate-sst and generate-plain metrics to measure snapshot generation speed. These metrics offer better visibility, allowing for a clearer comparison between snapshot generation throughput and send/receive throughput, making it easier to identify bottlenecks.
image
2.Add metrics for various snapshot tasks to the Snapshot Actions, including send, recv, recv_droppedn and others, providing better visibility for snapshot actions.
image

### What is changed and how it works?

Issue Number: Ref #15990

What's Changed:

```commit-message
This pull request refines metrics for the raft snapshot workflow, includes:
1. Use generate-kv, generate-sst and generate-plain metrics to measure snapshot generation speed. These metrics offer better visibility, allowing for a clearer comparison between snapshot generation throughput and send/receive throughput, making it easier to identify bottlenecks.
2. Add metrics for various snapshot tasks to the Snapshot Actions, including send, recv, recv_droppedn and others, providing better visibility for snapshot actions.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
none
```

### CODE REVIEW

File: components/raftstore/src/store/snap/io.rs [hunk 1]
```diff
@@ -39,7 +39,9 @@ pub trait StaleDetector {
 #[derive(Clone, Copy, Default)]
 pub struct BuildStatistics {
     pub key_count: usize,
-    pub total_size: usize,
+    pub total_kv_size: usize,
```
Comment (LykxSassinator) [inline components/raftstore/src/store/snap/io.rs:42]: Pls add some necessary annotations of the three different objects.

**Review by LykxSassinator — COMMENTED** [review]
overall LGTM

File: components/raftstore/src/store/snap/io.rs (hunk 1 above)
Comment (hbisheng) [reply components/raftstore/src/store/snap/io.rs:42]: +1, having some descriptions will be helpful. Is `total_kv_size` the uncompressed size and `total_sst_size` the compressed size?

File: components/raftstore/src/store/metrics.rs [hunk 2]
```diff
@@ -271,6 +271,12 @@ make_static_metric! {
         finished,
     }
 
+    pub label_enum SnapshotGeneraterBytesType {
```
[the remark this answers is not in the document]
Comment (hhwyt) [reply components/raftstore/src/store/metrics.rs]: done

File: components/raftstore/src/store/metrics.rs [hunk 3]
```diff
@@ -339,6 +345,10 @@ make_static_metric! {
             applystore_busy,
         },
     }
+
+    pub struct SnapshotGenerateBytesTypeVec: IntCounter {
+        "type" => SnapshotGeneraterBytesType,
```
[the remark this answers is not in the document]
Comment (hhwyt) [reply components/raftstore/src/store/metrics.rs]: done

File: components/raftstore/src/store/snap/io.rs (hunk 1 above)
Comment (hhwyt) [reply components/raftstore/src/store/snap/io.rs:48]: done

**Comment by hhwyt** [comment]
@hbisheng @LykxSassinator PTAL again, thx~  and if everything looks good, please give on ok-to-test.

File: components/raftstore/src/store/snap/io.rs (hunk 1 above)
Comment (hhwyt) [reply components/raftstore/src/store/snap/io.rs:48]: > +1, having some descriptions will be helpful. Is `total_kv_size` the uncompressed size and `total_sst_size` the compressed size?

yes.

**Review by LykxSassinator — APPROVED** [review]

**Review by hbisheng — APPROVED** [review]

**Pull request merged**